tx-pool/src/component/pending.rs

```diff
@@ -138,54 +164,71 @@ impl PendingQueue {
+        let removed = self.inner.remove(id);
 
-        if let Some(entry) = self.inner.remove(&id) {
-            for i in inputs {
-                self.inputs.remove(&i);
-            }
+        if let Some(ref entry) = removed {
+            self.remove_entry_relation(entry);
+        }
 
-            for d in related_out_points {
-                let mut empty = false;
-                if let Some(x) = self.deps.get_mut(d) {
-                    x.remove(&id);
-                    empty = x.is_empty();
-                }
+        removed
+    }
 
-                if empty {
-                    self.deps.remove(d);
+    pub(crate) fn remove_entry_and_descendants(&mut self, id: &ProposalShortId) -> Vec<TxEntry> {
+        let mut removed = Vec::new();
+        if let Some(entry) = self.inner.remove(id) {
+            let descendants = self.get_descendants(&entry);
+            self.remove_entry_relation(&entry);
+            removed.push(entry);
+            for id in descendants {
+                if let Some(entry) = self.remove_entry(&id) {
+                    removed.push(entry);
                 }
             }
-
-            self.header_deps.remove(&id);
-
-            return Some(entry);
         }
-        None
+        removed
     }
 
-    pub(crate) fn remove_entry(&mut self, id: &ProposalShortId) -> Option<TxEntry> {
-        let removed = self.inner.remove(id);
+    pub(crate) fn get_descendants(&self, entry: &TxEntry) -> Vec<ProposalShortId> {
+        use std::collections::VecDeque;
 
-        if let Some(ref entry) = removed {
-            self.remove_entry_relation(entry);
-        }
+        let mut entries: VecDeque<&TxEntry> = VecDeque::new();
+        entries.push_back(entry);
 
-        removed
+        let mut descendants = Vec::new();
+        while let Some(entry) = entries.pop_front() {
+            let outputs = entry.transaction().output_pts();
+
+            for output in outputs {
+                if let Some(ids) = self.outputs.get(&output) {
+                    descendants.extend(ids.iter().cloned());
+                    for id in ids {
+                        if let Some(entry) = self.inner.get(id) {
+                            entries.push_back(entry);
+                        }
+                    }
+                }
+            }
+        }
+        descendants
     }
 
     pub(crate) fn remove_entry_relation(&mut self, entry: &TxEntry) {
         let inputs = entry.transaction().input_pts_iter();
         let tx_short_id = entry.proposal_short_id();
+        let outputs = entry.transaction().output_pts();
 
         for i in inputs {
-            self.inputs.remove(&i);
+            let mut empty = false;
+
+            if let Some(ids) = self.inputs.get_mut(&i) {
+                ids.remove(&tx_short_id);
+                empty = ids.is_empty();
+            }
+
+            if empty {
+                self.inputs.remove(&i);
+            }
```

Comment:
suggest to use `Entry::Occupied`, which can reduce one hash lookup, this suggestion also apply to `// remove dep` code block

```suggestion
            if let Entry::Occupied(mut entry) = self.inputs.entry(i) {
                let empty = {
                    let ids = entry.get_mut();
                    ids.remove(&tx_short_id);
                    ids.is_empty()
                };
                if empty {
                    entry.remove();
                }
            }
```